File: src/test/java/cube8540/oauth/authentication/users/infra/DefaultUsernameValidationRuleTest.java
package cube8540.oauth.authentication.users.infra;

import cube8540.oauth.authentication.users.domain.User;
import cube8540.oauth.authentication.users.domain.Username;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@DisplayName("기본 유저 아이디 검사 테스트")
class DefaultUsernameValidationRuleTest {

    private static final Username USERNAME_GRATER_THEN_ALLOWED_LENGTH = new Username("username1234username1234username1234username1234");
    private static final Username USERNAME_LESS_THEN_ALLOWED_LENGTH = new Username("u12");
    private static final Username USERNAME_WITHOUT_NUMBER = new Username("username");
    private static final Username USERNAME_WITHOUT_CHARACTER = new Username("1234");
    private static final Username USERNAME_WITH_SPECIAL_CHARACTER = new Username("username!@#1234");
    private static final Username USERNAME = new Username("username1234");

    private DefaultUsernameValidationRule rule;

    @BeforeEach
    void setup() {
        this.rule = new DefaultUsernameValidationRule();
    }

    @Test
    @DisplayName("아이디가 허용 되는 문자 길이 보다 클시 유효성 검사 결과는 'false'가 반환 되어야 한다.")
    void usernameGraterThenAllowedLengthValidationResultShouldReturnFalse() {
        User user = mock(User.class);

        when(user.getUsername()).thenReturn(USERNAME_GRATER_THEN_ALLOWED_LENGTH);

        assertFalse(rule.isValid(user));
    }

    @Test
    @DisplayName("아이디가 허용 되는 문자 길이 보다 작을시 유효성 검사 결과는 'false'가 반환 되어야 한다.")
    void usernameLessThenAllowedLengthValidationResultShouldReturnFalse() {
        User user = mock(User.class);

        when(user.getUsername()).thenReturn(USERNAME_LESS_THEN_ALLOWED_LENGTH);

        assertFalse(rule.isValid(user));
    }

    @Test
    @DisplayName("아이디에 숫자가 포함 되지 않았을시 유효성 검사 결과는 'false'가 반환 되어야 한다.")
    void usernameNotContainsNumberValidationResultShouldReturnFalse() {
        User user = mock(User.class);

        when(user.getUsername()).thenReturn(USERNAME_WITHOUT_NUMBER);

        assertFalse(rule.isValid(user));
    }

    @Test
    @DisplayName("아이디에 문자가 포함 되지 않았을시 유효성 검사 결과는 'false'가 반환되여야 한다.")
    void usernameNotContainsCharacterValidationResultShouldReturnFalse() {
        User user = mock(User.class);

        when(user.getUsername()).thenReturn(USERNAME_WITHOUT_CHARACTER);

        assertFalse(rule.isValid(user));
    }

    @Test
    @DisplayName("아이디에 특수문자가 포함 되어 있을시 유효성 검사 결과는 'false'가 반횐 되어야 한다.")
    void usernameContainsSpacialCharacterValidationResultShouldReturnFalse() {
        User user = mock(User.class);

        when(user.getUsername()).thenReturn(USERNAME_WITH_SPECIAL_CHARACTER);

        assertFalse(rule.isValid(user));
    }

    @Test
    @DisplayName("아이디가 유효할시 유효성 검사 결과는 'true'가 반환 되어야 한다.")
    void usernameIsAllowedValidationResultShouldReturnTrue() {
        User user = mock(User.class);

        when(user.getUsername()).thenReturn(USERNAME);

        assertTrue(rule.isValid(user));
    }

}

File: src/test/kotlin/cube8540/oauth/authentication/user/infra/DefaultUsernameValidationRuleTest.kt
package cube8540.oauth.authentication.user.infra

import cube8540.oauth.authentication.users.domain.User
import cube8540.oauth.authentication.users.domain.Username
import cube8540.oauth.authentication.users.infra.DefaultUsernameValidationRule
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream


class DefaultUsernameValidationRuleTest {

    private val minUsernameLength = 4
    private val maxUsernameLength = 18
    private val charPool: List<Char> = ('a'..'z') + ('0'..'9')
    private val numberPool: IntRange = (0..9)
    private val characterPool: CharRange = ('a'..'z')

    private val rule: DefaultUsernameValidationRule = DefaultUsernameValidationRule()
    private val user: User = mockk()

    @ParameterizedTest
    @MethodSource(value = ["minLengthUsernameProvider", "maxLengthUsernameProvider"])
    fun `username invalid length`(username: Username) {
        every { user.username } returns username

        assertThat(rule.isValid(user)).isFalse
    }

    @ParameterizedTest
    @MethodSource(value = ["characterOnlyProvider"])
    fun `username exclude number`(username: Username) {
        every { user.username } returns username

        assertThat(rule.isValid(user)).isFalse
    }

    @ParameterizedTest
    @MethodSource(value = ["numberOnlyProvider"])
    fun `username exclude character`(username: Username) {
        every { user.username } returns username

        assertThat(rule.isValid(user)).isFalse
    }

    @Test
    fun `username included spacial character`() {
        every { user.username } returns Username("username!@#1234")

        assertThat(rule.isValid(user)).isFalse
    }

    @Test
    fun `username allowed`() {
        every { user.username } returns Username("username1234")

        assertThat(rule.isValid(user)).isTrue
    }

    private fun minLengthUsernameProvider(): Stream<Arguments> {
        val results = ArrayList<Username>()
        for (i in 1 until minUsernameLength) {
            val randomUsername = (1..i)
                .map { charPool.random() }
                .joinToString("")
            results.add(Username(randomUsername))
        }
        return results.map { Arguments.of(it) }.stream()
    }

    private fun maxLengthUsernameProvider(): Stream<Arguments> {
        val randomUsername = (1..(maxUsernameLength + 1))
            .map { charPool.random() }
            .joinToString("")
        return Stream.of(Arguments.of(Username(randomUsername)))
    }

    private fun characterOnlyProvider(): Stream<Arguments> {
        val results = ArrayList<Username>()
        for (i in minUsernameLength..maxUsernameLength) {
            val randomUsername = (0..i)
                .map { characterPool.random() }
                .joinToString("")
            results.add(Username(randomUsername))
        }
        return results.map { Arguments.of(it) }.stream()
    }

    private fun numberOnlyProvider(): Stream<Arguments> {
        val results = ArrayList<Username>()
        for (i in minUsernameLength..maxUsernameLength) {
            val randomUsername = (0..i)
                .map { numberPool.random() }
                .joinToString("")
            results.add(Username(randomUsername))
        }
        return results.map { Arguments.of(it) }.stream()
    }
}